src/FieldWarning/Assets/Ingame/UI/InputManager.cs

```diff
@@ -30,388 +32,327 @@ namespace PFW.Ingame.UI
-            if (_primedReticle == null)
-                throw new Exception("No primed reticle specified!");
-
-            _rightClickManager = new ClickManager(1, MoveGhostsToMouse, OnOrderShortClick, OnOrderLongClick, OnOrderHold);
+            Clicked,
+            Down,
+            Released
         }
 
-        void Update()
+        public enum Modes
         {
-            _selectionManager.Update(CurMouseMode);
-
-            switch (CurMouseMode) {
-
-            case MouseMode.purchasing:
-
-                RaycastHit hit;
-                if (Util.GetTerrainClickLocation(out hit)
-                    && hit.transform.gameObject.name.Equals("Terrain")) {
-
-                    ShowGhostUnitsAndMaybePurchase(hit);
-                }
-
-                MaybeExitPurchasingMode();
-                break;
-
-            case MouseMode.normal:
-                ApplyHotkeys();
-                _rightClickManager.Update();
-                break;
-
-            case MouseMode.firePos:
-                ApplyHotkeys();
-
-                if (Input.GetMouseButtonDown(0))
-                    _selectionManager.DispatchFirePosCommand();
-
-                if ((Input.GetMouseButtonDown(0) && !Input.GetKey(KeyCode.LeftShift))
-                    || Input.GetMouseButtonDown(1))
-                    EnterNormalMode();
-
-                break;
-
-            case MouseMode.reverseMove:
-                ApplyHotkeys();
-                if (Input.GetMouseButtonDown(0)) {
-                    MoveGhostsToMouse();
-                    _selectionManager.DispatchMoveCommand(
-                            false, MoveWaypoint.MoveMode.reverseMove);
-                }
-
-                if ((Input.GetMouseButtonDown(0) && !Input.GetKey(KeyCode.LeftShift))
-                    || Input.GetMouseButtonDown(1))
-                    EnterNormalMode();
-                break;
-
-            case MouseMode.fastMove:
-                ApplyHotkeys();
-                if (Input.GetMouseButtonDown(0)) {
-                    MoveGhostsToMouse();
-                    _selectionManager.DispatchMoveCommand(
-                            false, MoveWaypoint.MoveMode.fastMove);
-                }
+            Normal,
+            Purchasing,
+            FirePosition,
+            ReverseMove,
+            FastMove
+        }
 
-                if ((Input.GetMouseButtonDown(0) && !Input.GetKey(KeyCode.LeftShift)) 
-                    || Input.GetMouseButtonDown(1))
-                    EnterNormalMode();
-                break;
+        public Modes MouseState { get; private set; } = Modes.Normal;
 
-            default:
-                throw new Exception("impossible state");
-            }
+        public MatchSession Session
+        {
+            get => this._session;
+            set => this._session = this._session ?? value;
         }
 
-        public void OnGUI()
+        private PlayerData LocalPlayer => this.Session.LocalPlayer.Data;
```

Comment:
This is pretty neat and an interesting case of public/private naming. I guess keeping it private but named as a public does make the most sense.